sdks/go/pkg/beam/core/runtime/harness/harness.go

```diff
@@ -188,14 +198,27 @@ func (c *control) handleInstruction(ctx context.Context, req *fnpb.InstructionRe
 		log.Debugf(ctx, "PB: %v", msg)
 
 		ref := msg.GetProcessBundleDescriptorReference()
+		c.mu.Lock()
 		plan, ok := c.plans[ref]
+		// Make the plan active, and remove it from candidates
+		// since a plan can't be run concurrently.
```

Comment:
<!--thread_id:cc_161906666_t; commit:e1da5f2a09fbe4caf5901a1b3d000234fd49582d; resolved:1-->
<!--section:context-quote-->
> **wcn3** wrote:
> Currently, the implementation of the struct. It is more than a plan because it contains stateful elements that track work execution, like the source that I updated, as an example. The struct does allow the plan instance to be reused by resetting its state, but the fact that it has this state precludes concurrency.
> 
> There's already a JIRA to track the work needed to make the Go runner concurrent, and this implementation would need to be revisited at this time. The plan instances can (and already are) reused successfully, but this extra code was just a safety precaution in case the calling harness did something we didn't expect.

<!--section:body-->
Acknowledged.